In [ ]:
import contextkit.read as rd

In [ ]:
pdf= rd.read_pdf('FU-Technical Assessment - Applied Data Scientist-250425-033739.pdf')

In [ ]:
pdf

'Technical Assessment - Applied Data Scientist\nData Science Take-Home Challenge: Credit Risk Challenger Model Development\nIntroduction\nThank you for your interest in our data science position. This take-home challenge is designed to assess your abilities in developing and  \nvalidating credit risk models, specifically focusing on probability of default (PD) prediction. The challenge simulates a real-world scenario  \nwhere an existing model (Champion) needs improvement through the development of a new model (Challenger).\nIn the financial industry , having robust and accurate credit risk models is critical for making sound lending decisions and meeting  \nregulatory requirements. Your task is to build a challenger model that outperforms our current champion model in predicting the probability  \nof default for a loan portfolio.\nContext \nImagine a lender called AssetEase, a specialist in providing personal loans tailored for individuals looking to purchase significant assets  \nsuc

In [ ]:
from utils import *

In [ ]:
intro()

Note to solveit: Hi, I am Gaurav, and I am working with you today! When writing code for me, please follow fast.ai coding style: avoid comments; place single line if/loop/etc constructs on the same line; use concise variable names where appropriate. Follow the style you will see in my messages. As we work on this dialogue together.


Note to solveit: Hi, I am Gaurav, and I am working with you today! to solve this assessment  When writing code for me, please follow fast.ai coding style: avoid comments; place single line if/loop/etc constructs on the same line; use concise variable names where appropriate. Follow the style you will see in my messages. As we work on this dialogue together.

Now, can you walk me through assessment modeling Assessment Modeling Dataset.csv ? I want to understand what we're looking at and how we can use this dataset to challenge the champion model.


##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Looking at the Assessment Modeling Dataset, I can see it contains loan application data with various features and a target variable. Let's break it down:

The dataset contains information about loan applicants including:

1. **Borrower Information**:
   - `APP_ID`: Unique identifier for each application
   - `emp_title`: Employer name
   - `emp_length`: Years at current employer
   - `home_ownership`: Housing status (RENT, MORTGAGE, etc.)
   - `annual_inc`: Annual income
   - `addr_state`: State of residence
   - `zip_code`: ZIP code (partially masked)

2. **Loan Details**:
   - `purpose_cat` and `purpose`: Categorized and specific loan purpose
   - `Notes`: Text description provided by borrower
   - `verification_status`: Whether income was verified
   - `pymnt_plan`: Payment plan indicator

3. **Credit History**:
   - `debt_to_income`: Debt-to-income ratio
   - `delinq_2yrs`: Number of delinquencies in past 2 years
   - `earliest_cr_line`: Date of earliest credit line
   - `inq_last_6mths`: Number of inquiries in last 6 months
   - `open_acc`: Number of open credit accounts
   - `revol_bal`: Revolving balance
   - `revol_util`: Revolving line utilization rate
   - Various other credit metrics

4. **Target Variable**:
   - `TARGET`: Binary indicator (0 = no default, 1 = default)

This dataset will be used to:
1. Train your challenger model to predict the probability of default
2. Identify patterns and relationships between features and default outcomes
3. Create feature engineering strategies to improve prediction accuracy
4. Validate your model's performance before testing on the holdout set

To challenge the champion model effectively, you'll need to:
- Perform thorough exploratory data analysis to understand feature distributions
- Handle missing values (I can see some `nan` values in the data)
- Engineer new features that might better predict defaults
- Select appropriate algorithms and tune hyperparameters
- Validate using the Gini coefficient and other metrics mentioned in the assessment

Would you like me to help you start with some exploratory data analysis on this dataset?

In [ ]:
df = pd.read_csv('Assessment Modeling Dataset.csv')

In [ ]:
df.describe(include="object")

emp_title emp_length  ... initial_list_status policy_code
count       7535       8000  ...                8000        8000
unique      6688         14  ...                   2           5
top      US Army         10  ...                   f         PC3
freq          29       1717  ...                7985        1666

[4 rows x 13 columns]

In [ ]:
df.describe(include="number")

APP_ID     annual_inc  ...  mths_since_last_major_derog       TARGET
count   8000.000000    7999.000000  ...                  8000.000000  8000.000000
mean    5012.506875   68496.262437  ...                     2.004125     0.127750
std     2887.649416   49335.288010  ...                     0.811289     0.333832
min        2.000000    2000.000000  ...                     1.000000     0.000000
25%     2512.750000   40000.000000  ...                     1.000000     0.000000
50%     5014.500000   58344.000000  ...                     2.000000     0.000000
75%     7505.250000   82000.000000  ...                     3.000000     0.000000
max    10000.000000  900000.000000  ...                     3.000000     1.000000

[8 rows x 15 columns]

Based on your dataset exploration, I can see several columns with missing values. Here's a strategy for handling them:

1. **Low Missing Rate Columns (< 1%)**:
   - `annual_inc`, `delinq_2yrs`, `earliest_cr_line`, `inq_last_6mths`, `open_acc`, `pub_rec`, `total_acc`, `revol_util`, `collections_12_mths_ex_med`, `purpose`
   - Strategy: Impute with median (for numeric) or mode (for categorical)

2. **Medium Missing Rate Columns (1-40%)**:
   - `emp_title` (5.8%)
   - `Notes` (31.7%)
   - Strategy: For `emp_title`, create a "Unknown" category. For `Notes`, either drop the column if it's not predictive or use NLP techniques if you want to extract value

3. **High Missing Rate Columns (> 40%)**:
   - `mths_since_last_delinq` (63.4%)
   - `mths_since_last_record` (91.8%)
   - Strategy: Create binary features indicating whether the value is missing (often missing means "no delinquency" or "no record"), then impute the non-missing values

Here's code to implement this approach:

```python
# For low missing rate numeric columns
num_cols = ['annual_inc', 'delinq_2yrs', 'inq_last_6mths', 'open_acc', 
           'pub_rec', 'total_acc', 'revol_util', 'collections_12_mths_ex_med']
for col in num_cols:
    if df[col].isna().sum() > 0:
        df[col] = df[col].fillna(df[col].median())

# For categorical columns with low missing rates
cat_cols = ['purpose', 'earliest_cr_line']
for col in cat_cols:
    if df[col].isna().sum() > 0:
        df[col] = df[col].fillna(df[col].mode()[0])

# For medium missing rate columns
if df['emp_title'].isna().sum() > 0:
    df['emp_title'] = df['emp_title'].fillna('Unknown')

# For high missing rate columns, create indicator features
for col in ['mths_since_last_delinq', 'mths_since_last_record']:
    df[f'has_{col}'] = df[col].notna().astype(int)
    df[col] = df[col].fillna(0)  # Missing often means "no delinquency/record"
```

In [ ]:
low_missing_cols = missing_cols[missing_cols< 0.01].index

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 low_missing_cols = missing_cols[missing_cols< 0.01].index

NameError: name 'missing_cols' is not defined


NameError: name 'missing_cols' is not defined

In [ ]:
low_missing_cols = missing_cols[(missing_cols< 0.01) & (missing_cols>0)]
low_missing_cols


---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 low_missing_cols = missing_cols[(missing_cols< 0.01) & (missing_cols>0)]
      2 low_missing_cols

NameError: name 'missing_cols' is not defined


NameError: name 'missing_cols' is not defined

In [ ]:
mode_values = {}
for col in low_missing_cols.index:
    mode_values[col] = df[col].mode()[0]
mode_values

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 2
      1 mode_values = {}
----> 2 for col in low_missing_cols.index:
      3     mode_values[col] = df[col].mode()[0]
      4 mode_values

NameError: name 'low_missing_cols' is not defined


NameError: name 'low_missing_cols' is not defined

## replace wit with mode vaue for 0.01 missing value

In [ ]:
for col in low_missing_cols.index:
    df[col] = df[col].fillna(mode_values[col])

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 for col in low_missing_cols.index:
      2     df[col] = df[col].fillna(mode_values[col])

NameError: name 'low_missing_cols' is not defined


NameError: name 'low_missing_cols' is not defined

In [ ]:
low_missing_cols = missing_cols[(missing_cols< 0.01) & (missing_cols>0)]

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 low_missing_cols = missing_cols[(missing_cols< 0.01) & (missing_cols>0)]

NameError: name 'missing_cols' is not defined


NameError: name 'missing_cols' is not defined

In [ ]:
df[low_missing_cols.index].isna().sum()

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 df[low_missing_cols.index].isna().sum()

NameError: name 'low_missing_cols' is not defined


NameError: name 'low_missing_cols' is not defined

## Medium Missing Rate Columns (1-40%)

In [ ]:
#lets get the medium missing rate columns (1-40%)
missing_cols=    df.isna().sum()/df.shape[0]
medium_missing_cols = missing_cols[(missing_cols< 0.40) & (missing_cols>0.01)]
medium_missing_cols





emp_title    0.058125
Notes        0.316500
dtype: float64

In [ ]:
medium_missing_cols

emp_title    0.058125
Notes        0.316500
dtype: float64

In [ ]:
# df[medium_missing_cols.index]
mode_values = {}
for col in medium_missing_cols.index:
    mode_values[col] = df[col].mode()[0]
mode_values

{'emp_title': 'US Army', 'Notes': ' '}

In [ ]:
df[medium_missing_cols.index].isna().sum()

emp_title     465
Notes        2532
dtype: int64

In [ ]:
df['emp_title'] = df['emp_title'].fillna('Unknown')

In [ ]:
df['Notes'] = df['Notes'].fillna('Not shared')

## High Missing Rate Columns (> 40%)

In [ ]:
# get more than 40 % columns 
missing_cols=    df.isna().sum()/df.shape[0]
high_missing_cols = missing_cols[(missing_cols> 0.40)]
high_missing_cols


mths_since_last_delinq    0.634375
mths_since_last_record    0.918000
dtype: float64

In [ ]:
mode_values = {}
for col in high_missing_cols.index:
    mode_values[col] = df[col].mode()[0]
mode_values

{'mths_since_last_delinq': np.float64(0.0),
 'mths_since_last_record': np.float64(0.0)}

In [ ]:
df[high_missing_cols.index]

mths_since_last_delinq  mths_since_last_record
0                        NaN                     NaN
1                       44.0                     NaN
2                        NaN                     NaN
3                       46.0                     NaN
4                       22.0                     NaN
...                      ...                     ...
7995                     NaN                     NaN
7996                     NaN                     NaN
7997                    49.0                     NaN
7998                    44.0                     NaN
7999                     NaN                     NaN

[8000 rows x 2 columns]

In [ ]:
for col in high_missing_cols.index:
    df[f'has_{col}'] = df[col].notna().astype(int)

In [ ]:
df.columns

Index(['APP_ID', 'emp_title', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'pymnt_plan', 'Notes', 'purpose_cat',
       'purpose', 'zip_code', 'addr_state', 'debt_to_income', 'delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths',
       'mths_since_last_delinq', 'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc',
       'initial_list_status', 'collections_12_mths_ex_med', 'mths_since_last_major_derog', 'policy_code', 'TARGET',
       'has_mths_since_last_delinq', 'has_mths_since_last_record'],
      dtype='object')

In [ ]:
df[['mths_since_last_delinq', 'has_mths_since_last_delinq']]
    

mths_since_last_delinq  has_mths_since_last_delinq
0                        NaN                           0
1                       44.0                           1
2                        NaN                           0
3                       46.0                           1
4                       22.0                           1
...                      ...                         ...
7995                     NaN                           0
7996                     NaN                           0
7997                    49.0                           1
7998                    44.0                           1
7999                     NaN                           0

[8000 rows x 2 columns]

In [ ]:
for col in high_missing_cols.index:
    df[col] = df[col].fillna(0)

In [ ]:
df[['mths_since_last_delinq', 'has_mths_since_last_delinq', 
    'mths_since_last_record', 'has_mths_since_last_record']].head(10)

mths_since_last_delinq  ...  has_mths_since_last_record
0                     0.0  ...                           0
1                    44.0  ...                           0
2                     0.0  ...                           0
3                    46.0  ...                           0
4                    22.0  ...                           0
5                     0.0  ...                           0
6                     0.0  ...                           0
7                     0.0  ...                           0
8                     0.0  ...                           0
9                     0.0  ...                           0

[10 rows x 4 columns]

In [ ]:
def preprocess_data(df):
    "Preprocess loan application data with appropriate handling of missing values"
    # Make a copy to avoid modifying original
    df = df.copy()
    
    # Calculate missing percentages
    missing_pct = df.isna().sum()/df.shape[0]
    
    # Handle low missing rate columns (<1%)
    low_missing = missing_pct[(missing_pct < 0.01) & (missing_pct > 0)].index
    for col in low_missing:
        mode_val = df[col].mode()[0]
        df[col] = df[col].fillna(mode_val)
    
    # Handle medium missing rate columns (1-40%)
    df['emp_title'] = df['emp_title'].fillna('Unknown')
    df['Notes'] = df['Notes'].fillna('Not shared')
    
    # Handle high missing rate columns (>40%)
    high_missing = missing_pct[missing_pct > 0.40].index
    for col in high_missing:
        df[f'has_{col}'] = df[col].notna().astype(int)
        df[col] = df[col].fillna(0)
    
    return df


In [ ]:
df.isna().sum()

APP_ID                          0
emp_title                       0
emp_length                      0
home_ownership                  0
annual_inc                      1
verification_status             0
pymnt_plan                      0
Notes                           0
purpose_cat                     0
purpose                         4
zip_code                        0
addr_state                      0
debt_to_income                  0
delinq_2yrs                     5
earliest_cr_line                5
inq_last_6mths                  5
mths_since_last_delinq          0
mths_since_last_record          0
open_acc                        5
pub_rec                         5
revol_bal                       0
revol_util                     23
total_acc                       5
initial_list_status             0
collections_12_mths_ex_med     29
mths_since_last_major_derog     0
policy_code                     0
TARGET                          0
has_mths_since_last_delinq      0
has_mths_since

In [ ]:
df = preprocess_data(df)

In [ ]:
from utils import *

# Faeture eng for ltr
```python
# Check target distribution
df['TARGET'].value_counts(normalize=True)
```

Then let's create some useful features that might help predict defaults:

```python
# Extract credit history length
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'])
df['cr_history_length'] = (pd.to_datetime('now') - df['earliest_cr_line']).dt.days / 365.25

# Create debt-to-income ratio buckets
df['dti_buckets'] = pd.cut(df['debt_to_income'], bins=[0, 10, 15, 20, 30], labels=['Low', 'Medium', 'High', 'Very High'])

# Create utilization bands
df['util_band'] = pd.cut(df['revol_util'], bins=[0, 30, 60, 80, 100], labels=['Low', 'Medium', 'High', 'Very High'])

# Create inquiry bands
df['inq_band'] = pd.cut(df['inq_last_6mths'], bins=[-1, 0, 2, 5, 25], labels=['None', 'Low', 'Medium', 'High'])
```

Next, let's look at correlation with the target:

```python
# Calculate correlation with target for numeric features
num_cols = df.select_dtypes(include=['number']).columns
correlations = df[num_cols].corrwith(df['TARGET']).sort_values(ascending=False)
correlations
```

This will give us a good foundation for building our challenger model.

In [ ]:
df.TARGET.value_counts(normalize=True)

TARGET
0    0.87225
1    0.12775
Name: proportion, dtype: float64

lest go with class weight

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Let's implement the tabular model with class weights to handle the imbalance:

```python
from fastai.tabular.all import *

# Split data into train and validation sets
train_idx = np.random.permutation(len(df))[:int(0.8*len(df))]
valid_idx = np.setdiff1d(np.arange(len(df)), train_idx)

# Define categorical and continuous variables
cat_names = ['home_ownership', 'verification_status', 'purpose_cat', 'addr_state', 
             'initial_list_status', 'policy_code', 'has_mths_since_last_delinq', 
             'has_mths_since_last_record']
             
cont_names = ['annual_inc', 'debt_to_income', 'delinq_2yrs', 'inq_last_6mths', 
              'mths_since_last_delinq', 'mths_since_last_record', 'open_acc', 
              'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 
              'mths_since_last_major_derog']

# Create TabularPandas object
to = TabularPandas(df, procs=[Categorify, FillMissing, Normalize],
                  cat_names=cat_names,
                  cont_names=cont_names,
                  y_names='TARGET',
                  splits=(train_idx, valid_idx))

# Create DataLoaders
dls = to.dataloaders(bs=64)

# Calculate class weights (inverse of class frequencies)
weights = 1/torch.Tensor([0.87225, 0.12775])
weights = weights / weights.sum()
class_weights = weights * 2  # Normalize and scale

# Create and train the model with class weights
learn = tabular_learner(dls, metrics=[accuracy, RocAucBinary()])
learn.loss_func = CrossEntropyLossFlat(weight=class_weights)

# Train the model
learn.fit_one_cycle(5, 1e-2, wd=0.1)
```

what about the the Gini coefficient metric..

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

You're right, we should include the Gini coefficient since it's specifically mentioned in your assessment. Let's implement a custom Gini metric for fastai:

```python
class GiniCoefficient(Metric):
    "Computes Gini coefficient (2*AUC-1)"
    def __init__(self): self.reset()
    
    def reset(self): 
        self.preds,self.targs = [],[]
        
    def accumulate(self, learn):
        pred = learn.pred.sigmoid()
        self.preds.append(pred)
        self.targs.append(learn.y)
    
    @property
    def value(self):
        preds = torch.cat(self.preds).cpu().numpy()
        targs = torch.cat(self.targs).cpu().numpy()
        auc = roc_auc_score(targs, preds)
        return 2 * auc - 1
    
    @property
    def name(self): return "gini"
```

Now let's update our model to include this metric:

```python
# Create and train the model with class weights and Gini coefficient
learn = tabular_learner(dls, metrics=[accuracy, RocAucBinary(), GiniCoefficient()])
learn.loss_func = CrossEntropyLossFlat(weight=class_weights)

# Train the model
learn.fit_one_cycle(5, 1e-2, wd=0.1)
```

In [ ]:
cont_cols,cat_cols = cont_cat_split(df,dep_var='TARGET')

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cont_cols,cat_cols = cont_cat_split(df,dep_var='TARGET')

NameError: name 'cont_cat_split' is not defined


NameError: name 'cont_cat_split' is not defined

In [ ]:
cont_cols

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cont_cols

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

In [ ]:
df[cont_cols].head()

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 df[cont_cols].head()

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

In [ ]:
df[cat_cols].head()

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 df[cat_cols].head()

NameError: name 'cat_cols' is not defined


NameError: name 'cat_cols' is not defined

In [ ]:
#use RandomSplitter to get 80 train and 20 validation data
splits = RandomSplitter(valid_pct=0.2, seed=42)(range_of(df))

In [ ]:
splits

((#6400) [np.int64(1449),np.int64(4488),np.int64(2985),np.int64(5470),np.int64(4407),np.int64(6658),np.int64(5757),np.int64(2280),np.int64(4869),np.int64(3909),np.int64(6662),np.int64(4173),np.int64(520),np.int64(1408),np.int64(3099),np.int64(936),np.int64(1613),np.int64(4317),np.int64(2253),np.int64(3210)...],
 (#1600) [np.int64(5542),np.int64(1762),np.int64(3954),np.int64(5968),np.int64(6758),np.int64(1615),np.int64(4484),np.int64(915),np.int64(5718),np.int64(1738),np.int64(5918),np.int64(4242),np.int64(1982),np.int64(5421),np.int64(7991),np.int64(7997),np.int64(7203),np.int64(3000),np.int64(6585),np.int64(7931)...])

In [ ]:
df[cat_cols].nunique()

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 df[cat_cols].nunique()

NameError: name 'cat_cols' is not defined


NameError: name 'cat_cols' is not defined

In [ ]:
def cat_summary(df, cat_cols, top_n=5):
    "Create a clean summary of categorical variables"
    results = {}
    
    for col in cat_cols:
        # Get value counts and calculate percentages
        counts = df[col].value_counts()
        total = len(df)
        pcts = counts / total * 100
        
        # Store top N categories and their percentages
        top_cats = list(counts.index[:top_n])
        top_pcts = list(pcts.values[:top_n])
        
        # Format percentages with 2 decimal places
        formatted_pcts = [f"{p:.2f}%" for p in top_pcts]
        
        # Create a dict of category:percentage pairs
        cat_pct_pairs = dict(zip(top_cats, formatted_pcts))
        
        # Add number of unique values
        n_unique = df[col].nunique()
        
        results[col] = {
            'unique_values': n_unique,
            'top_categories': cat_pct_pairs
        }
    
    return results

# Generate the summary
summary = cat_summary(df, cat_cols)

# Display in a more readable format
for col, data in summary.items():
    print(f"\n{col} ({data['unique_values']} unique values)")
    print("-" * 50)
    for cat, pct in data['top_categories'].items():
        print(f"{cat:<30} {pct:>10}")


---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 32
     29     return results
     31 # Generate the summary
---> 32 summary = cat_summary(df, cat_cols)
     34 # Display in a more readable format
     35 for col, data in summary.items():

NameError: name 'cat_cols' is not defined


NameError: name 'cat_cols' is not defined

what's the ideal strategy for handling high cardinality categorical values? Specifically, "notes" shouldn't be treated as categorical. "Employee title" has really high cardinality, so I think we should drop it and convert the  continuous col. Definitely drop "notes" and "purpose." Also, "zip code." And "earliest credit line" is a date format, so we need to pull that out of the categorical values and handle it properly. please share your opinions ...


##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

You're absolutely right about handling these high cardinality features. Here's my recommended approach:

For high cardinality categorical columns:

1. **emp_title (6689 unique values)**: 
   - Drop this column as it has too many unique values to be useful as a categorical
   - Alternatively, we could extract a binary feature like "is_employed" but since most applicants are employed, this likely adds little value

2. **Notes (5411 unique values)** and **purpose (4678 unique values)**:
   - Drop these columns as they're too granular
   - Keep only `purpose_cat` which has a manageable 27 categories and captures the same information in a structured way

3. **zip_code (688 unique values)**:
   - Drop this as it's too granular
   - We already have `addr_state` which provides geographic information at a more appropriate level

4. **earliest_cr_line (447 unique values)**:
   - Convert this date to useful numeric features:
   - Credit history length in years
   - Possibly extract month as seasonality feature

Here's how I'd implement these changes:

```python
# Drop high cardinality and less useful columns
cols_to_drop = ['emp_title', 'Notes', 'purpose', 'zip_code']
df_clean = df.drop(columns=cols_to_drop)

# Convert earliest_cr_line to credit history length
df_clean['earliest_cr_line'] = pd.to_datetime(df_clean['earliest_cr_line'])
today = pd.to_datetime('today')
df_clean['credit_history_years'] = (today - df_clean['earliest_cr_line']).dt.days / 365.25

# Drop the original date column after extracting features
df_clean = df_clean.drop(columns=['earliest_cr_line'])

# Update categorical columns list
cat_cols = [col for col in cat_cols if col not in cols_to_drop + ['earliest_cr_line']]

# Add the new numeric column to continuous columns
cont_cols = cont_cols + ['credit_history_years']
```

This approach will significantly reduce dimensionality while preserving the most valuable information for your model.

In [ ]:
# lets drop for now(we ll come back to this ltr)
# Drop high cardinality and less useful columns
cols_to_drop = ['emp_title', 'Notes', 'purpose', 'zip_code']
df = df.drop(columns=cols_to_drop)



In [ ]:
df.columns

Index(['APP_ID', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'pymnt_plan', 'purpose_cat', 'addr_state',
       'debt_to_income', 'delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record',
       'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'initial_list_status', 'collections_12_mths_ex_med',
       'mths_since_last_major_derog', 'policy_code', 'TARGET', 'has_mths_since_last_delinq', 'has_mths_since_last_record'],
      dtype='object')

In [ ]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'])

In [ ]:
df['earliest_cr_line'].dt.year

0       1999
1       1996
2       1997
3       2003
4       2001
        ... 
7995    2002
7996    1982
7997    2003
7998    1996
7999    1994
Name: earliest_cr_line, Length: 8000, dtype: int32

In [ ]:
df['earliest_cr_line'].dt.year.value_counts()

earliest_cr_line
2000    666
1998    610
1999    558
2001    512
1997    483
1996    480
1995    423
1994    413
2002    410
2003    378
2004    337
2005    323
1993    296
2006    256
1991    213
1990    210
1992    195
1989    168
2007    138
1987    138
1988    128
1986     98
1984     91
1985     71
1983     59
1982     50
1979     39
1981     33
1980     32
1978     31
2008     27
1976     22
1977     21
1975     14
1974     11
1973     11
1970     10
1972      9
1971      8
1969      7
1968      5
1967      4
1965      2
1964      2
1963      2
1966      2
1956      1
1962      1
1950      1
1959      1
Name: count, dtype: int64

how How should we handle this date column? The maximum year we have is 2000, so putting the current year, 2024, doesn't make sense. What's the best strategy to handle this column and get some meaningful information out of it?


##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Looking at the distribution, the credit report years range from 1950 to 2008, with most concentrated in the 1990s and early 2000s. Since this appears to be historical data, we should calculate credit age relative to a reference date that makes sense for this dataset rather than using today's date.

Here are better approaches:

1. Calculate credit history length relative to a reference date in the dataset:

```python
# Use the maximum date in the dataset as reference point
max_date = df_clean['earliest_cr_line'].max()
print(f"Maximum credit line date: {max_date}")

# Calculate credit history length relative to this date
df_clean['credit_history_years'] = (max_date - df_clean['earliest_cr_line']).dt.days / 365.25
```

2. Create age buckets based on the distribution:

```python
# Create credit history age bins
df_clean['cr_age_bucket'] = pd.cut(
    df_clean['earliest_cr_line'].dt.year,
    bins=[1950, 1980, 1990, 1995, 2000, 2005, 2010],
    labels=['Pre-1980', '1980s', 'Early 90s', 'Late 90s', 'Early 00s', 'Mid 00s']
)
```

3. Calculate years since first credit:

```python
# Simple years since first credit (relative to 2008)
reference_year = 2008
df_clean['years_since_first_cr'] = reference_year - df_clean['earliest_cr_line'].dt.year
```

These approaches will give you meaningful features about credit history length without using today's date.

I agree lets do that...

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Let's implement the credit history years calculation using the reference year approach:

```python
# Use 2008 as reference year (latest in dataset)
reference_year = 2008
df_clean['years_since_first_cr'] = reference_year - df_clean['earliest_cr_line'].dt.year

# Drop the original datetime column since we've extracted what we need
df_clean = df_clean.drop(columns=['earliest_cr_line'])

# Add the new feature to continuous columns
cont_cols = [c for c in cont_cols if c != 'earliest_cr_line'] + ['years_since_first_cr']

# Quick check of the new feature
df_clean['years_since_first_cr'].describe()
```

In [ ]:
max_date = df['earliest_cr_line'].max()
print(f"Maximum credit line date: {max_date}")

Maximum credit line date: 2008-11-01 00:00:00


In [ ]:
df['years_since_first_cr'] = 2008 - df['earliest_cr_line'].dt.year

In [ ]:
df['years_since_first_cr']

0        9
1       12
2       11
3        5
4        7
        ..
7995     6
7996    26
7997     5
7998    12
7999    14
Name: years_since_first_cr, Length: 8000, dtype: int32

In [ ]:
cont_cols = [c for c in cont_cols if c != 'earliest_cr_line'] + ['years_since_first_cr']

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cont_cols = [c for c in cont_cols if c != 'earliest_cr_line'] + ['years_since_first_cr']

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

In [ ]:
cont_cols

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cont_cols

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

In [ ]:
cat_cols

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cat_cols

NameError: name 'cat_cols' is not defined


NameError: name 'cat_cols' is not defined

drop this from cat _cols (cols_to_drop = ['emp_title', 'Notes', 'purpose', 'zip_code'])

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Let's update the categorical columns list by removing the dropped columns:

```python
cols_to_drop = ['emp_title', 'Notes', 'purpose', 'zip_code', 'earliest_cr_line']
cat_cols = [col for col in cat_cols if col not in cols_to_drop]
cat_cols
```

In [ ]:
cols_to_drop = ['emp_title', 'Notes', 'purpose', 'zip_code', 'earliest_cr_line']
cat_cols = [col for col in cat_cols if col not in cols_to_drop]
cat_cols

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 2
      1 cols_to_drop = ['emp_title', 'Notes', 'purpose', 'zip_code', 'earliest_cr_line']
----> 2 cat_cols = [col for col in cat_cols if col not in cols_to_drop]
      3 cat_cols

NameError: name 'cat_cols' is not defined


NameError: name 'cat_cols' is not defined

let wrap above steps in concise function and keep preprocess_data as seprate function only do not combine that with new function i want to call them seprate on data..


##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's a concise function to handle feature engineering separately from the preprocessing:

```python
def feature_engineering(df):
    "Engineer new features and clean up existing ones"
    df = df.copy()
    
    # Drop high cardinality and less useful columns
    cols_to_drop = ['emp_title', 'Notes', 'purpose', 'zip_code']
    df = df.drop(columns=[c for c in cols_to_drop if c in df.columns])
    
    # Convert earliest_cr_line to credit history years
    if 'earliest_cr_line' in df.columns:
        df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'])
        reference_year = 2008
        df['years_since_first_cr'] = reference_year - df['earliest_cr_line'].dt.year
        df = df.drop(columns=['earliest_cr_line'])
    
    return df
```

This keeps it separate from your existing `preprocess_data` function and focuses only on feature engineering steps.

pefect Remember, I don't want to merge everything. Keep the functions concise and separate. What you've done right now is perfect.


##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great, I will remember that...



In [ ]:
def feature_engineering(df):
    "Engineer new features and clean up existing ones"
    df = df.copy()
    
    # Drop high cardinality and less useful columns
    cols_to_drop = ['emp_title', 'Notes', 'purpose', 'zip_code','APP_ID','collections_12_mths_ex_med'] # collections_12_mths_ex_med i unique value and APP_ID(all unique)
    df = df.drop(columns=[c for c in cols_to_drop if c in df.columns])
    
    # Convert earliest_cr_line to credit history years
    if 'earliest_cr_line' in df.columns:
        df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'])
        reference_year = 2008
        df['years_since_first_cr'] = reference_year - df['earliest_cr_line'].dt.year
        df = df.drop(columns=['earliest_cr_line'])
    
    return df

In [ ]:
df= feature_engineering(df)

In [ ]:
df

emp_length  ... years_since_first_cr
0             9  ...                    9
1             2  ...                   12
2            10  ...                   11
3             3  ...                    5
4             1  ...                    7
...         ...  ...                  ...
7995          1  ...                    6
7996          5  ...                   26
7997          5  ...                    5
7998          5  ...                   12
7999          6  ...                   14

[8000 rows x 24 columns]

In [ ]:
cont_cat_split?

Object `cont_cat_split` not found.


In [ ]:
cont_cols,cat_cols = cont_cat_split(df,dep_var='TARGET')

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cont_cols,cat_cols = cont_cat_split(df,dep_var='TARGET')

NameError: name 'cont_cat_split' is not defined


NameError: name 'cont_cat_split' is not defined

In [ ]:
cont_cols

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cont_cols

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

In [ ]:
cat_cols

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cat_cols

NameError: name 'cat_cols' is not defined


NameError: name 'cat_cols' is not defined

In [ ]:
df[cont_cols]

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 df[cont_cols]

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

In [ ]:
df[cont_cols].isna().sum()

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 df[cont_cols].isna().sum()

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

In [ ]:
df[cont_cols].nunique()

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 df[cont_cols].nunique()

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

In [ ]:

df[cont_cols].describe()

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 df[cont_cols].describe()

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

In [ ]:
len(df['APP_ID'].unique())

---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File /usr/local/lib/python3.11/site-packages/pandas/core/indexes/base.py:3805, in Index.get_loc(self, key)
   3804 try:
-> 3805     return self._engine.get_loc(casted_key)
   3806 except KeyError as err:

File index.pyx:167, in pandas._libs.index.IndexEngine.get_loc()

File index.pyx:196, in pandas._libs.index.IndexEngine.get_loc()

File pandas/_libs/hashtable_class_helper.pxi:7081, in pandas._libs.hashtable.PyObjectHashTable.get_item()

File pandas/_libs/hashtable_class_helper.pxi:7089, in pandas._libs.hashtable.PyObjectHashTable.get_item()

KeyError: 'APP_ID'

The above exception was the direct cause of the following exception:

KeyError                                  Traceback (most recent call last)
Cell In[1], line 1
----> 1 len(df['APP_ID'].unique())

File /usr/local/lib/python3.11/site-packages/pandas/core/frame.py:4102, in Dat

KeyError: 'APP_ID'

In [ ]:
df['pub_rec'].value_counts()

pub_rec
0.0    7562
1.0     419
2.0      14
3.0       5
Name: count, dtype: int64

For columns like `delinq_2yrs`, `pub_rec`, `collections_12`, `open_acc_6m`, and `inq_last_6m`, the cardinality is too low. For example, `delinq_2yrs` represents delinquencies in the last two years, `pub_rec` is the number of public records, and `collections_12` is the number of collections in the last 12 months.

Given the low cardinality, should we treat these as continuous or convert them to categorical features? What's the best strategy here?


##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

For low cardinality numeric features like these, there are two reasonable approaches:

1. Keep them as continuous features, which preserves their numerical relationships
2. Convert them to categorical features, which can capture non-linear patterns

Here's my recommendation for each column:

**Convert to categorical:**
- `pub_rec` (4 values): The difference between 0, 1, 2, and 3 public records may not be linear in terms of risk
- `collections_12_mths_ex_med` (1 value): This has only one value (0) so it's not providing any information - consider dropping it

**Keep as continuous:**
- `delinq_2yrs` (10 values): While it has low cardinality, there's likely a continuous relationship between number of delinquencies and default risk
- `inq_last_6mths` (18 values): The relationship between inquiries and risk is likely somewhat linear
- `open_acc` (35 values): This has enough values to be treated as continuous

For the ones we convert to categorical, we can do:

```python
# Convert selected low-cardinality features to categorical
for col in ['pub_rec']:
    df[f'{col}_cat'] = df[col].astype('category')
    cat_cols.append(f'{col}_cat')

# Drop collections_12_mths_ex_med since it has only one value
if 'collections_12_mths_ex_med' in cont_cols:
    cont_cols.remove('collections_12_mths_ex_med')
```

This balanced approach gives us the best of both worlds - preserving the numerical relationships where they make sense while allowing the model to learn non-linear patterns for features where the numerical relationship might not be strictly linear.

In [ ]:
df['collections_12_mths_ex_med'].value_counts()

---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File /usr/local/lib/python3.11/site-packages/pandas/core/indexes/base.py:3805, in Index.get_loc(self, key)
   3804 try:
-> 3805     return self._engine.get_loc(casted_key)
   3806 except KeyError as err:

File index.pyx:167, in pandas._libs.index.IndexEngine.get_loc()

File index.pyx:196, in pandas._libs.index.IndexEngine.get_loc()

File pandas/_libs/hashtable_class_helper.pxi:7081, in pandas._libs.hashtable.PyObjectHashTable.get_item()

File pandas/_libs/hashtable_class_helper.pxi:7089, in pandas._libs.hashtable.PyObjectHashTable.get_item()

KeyError: 'collections_12_mths_ex_med'

The above exception was the direct cause of the following exception:

KeyError                                  Traceback (most recent call last)
Cell In[1], line 1
----> 1 df['collections_12_mths_ex_med'].value_counts()

File /usr/local/lib/python3.11/site-p

KeyError: 'collections_12_mths_ex_med'

In [ ]:
df['pub_rec'].value_counts()

pub_rec
0.0    7562
1.0     419
2.0      14
3.0       5
Name: count, dtype: int64

In [ ]:
df['delinq_2yrs'].value_counts()

delinq_2yrs
0.0     7120
1.0      666
2.0      148
3.0       42
4.0       13
5.0        5
6.0        2
7.0        2
8.0        1
11.0       1
Name: count, dtype: int64

In [ ]:
df['inq_last_6mths'].value_counts()

inq_last_6mths
0.0     3690
1.0     2143
2.0     1134
3.0      592
4.0      181
5.0      123
6.0       60
7.0       34
8.0       22
9.0       10
18.0       2
10.0       2
11.0       2
16.0       1
25.0       1
24.0       1
14.0       1
12.0       1
Name: count, dtype: int64

In [ ]:

df['open_acc'].value_counts()

open_acc
7.0     830
6.0     789
8.0     745
9.0     742
10.0    633
5.0     612
11.0    554
4.0     509
12.0    474
13.0    396
3.0     294
14.0    280
15.0    252
16.0    180
17.0    144
2.0     125
18.0    107
19.0     85
20.0     61
21.0     58
22.0     29
23.0     28
25.0     19
24.0     12
26.0      7
1.0       7
27.0      6
30.0      6
28.0      5
29.0      3
31.0      2
33.0      2
36.0      2
35.0      1
32.0      1
Name: count, dtype: int64

In [ ]:
for col in ['pub_rec']:
    df[f'{col}_cat'] = df[col].astype('category')
    cat_cols.append(f'{col}_cat')

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 3
      1 for col in ['pub_rec']:
      2     df[f'{col}_cat'] = df[col].astype('category')
----> 3     cat_cols.append(f'{col}_cat')

NameError: name 'cat_cols' is not defined


NameError: name 'cat_cols' is not defined

In [ ]:
df

emp_length home_ownership  ...  years_since_first_cr pub_rec_cat
0             9       MORTGAGE  ...                     9         0.0
1             2           RENT  ...                    12         0.0
2            10       MORTGAGE  ...                    11         0.0
3             3           RENT  ...                     5         0.0
4             1       MORTGAGE  ...                     7         0.0
...         ...            ...  ...                   ...         ...
7995          1           RENT  ...                     6         0.0
7996          5       MORTGAGE  ...                    26         0.0
7997          5            OWN  ...                     5         0.0
7998          5           RENT  ...                    12         0.0
7999          6       MORTGAGE  ...                    14         0.0

[8000 rows x 25 columns]

In [ ]:
# df['pub_rec']

In [ ]:
def feature_engineering(df):
    "Engineer new features and clean up existing ones"
    df = df.copy()
    
    # Drop high cardinality and less useful columns
    cols_to_drop = ['emp_title', 'Notes', 'purpose', 'zip_code', 'APP_ID', 'collections_12_mths_ex_med'] 
    df = df.drop(columns=[c for c in cols_to_drop if c in df.columns])
    
    # Convert earliest_cr_line to credit history years
    if 'earliest_cr_line' in df.columns:
        df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'])
        reference_year = 2008
        df['years_since_first_cr'] = reference_year - df['earliest_cr_line'].dt.year
        df = df.drop(columns=['earliest_cr_line'])
    
    # Handle low cardinality features
    # Convert pub_rec to categorical (very skewed distribution, only 4 values)
    if 'pub_rec' in df.columns:
        df['pub_rec_cat'] = df['pub_rec'].astype('category')
    
    # We keep delinq_2yrs and inq_last_6mths as continuous since they have more values
    # and likely have a somewhat linear relationship with default risk
    
    return df

In [ ]:
df = feature_engineering(df)

In [ ]:
cont_cat_split?

Object `cont_cat_split` not found.


In [ ]:
cont_cols,cat_cols = cont_cat_split(df, max_card=20, dep_var='TARGET')


---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cont_cols,cat_cols = cont_cat_split(df, max_card=20, dep_var='TARGET')

NameError: name 'cont_cat_split' is not defined


NameError: name 'cont_cat_split' is not defined

In [ ]:
cont_cols

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cont_cols

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

In [ ]:
df[cont_cols].nunique()

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 df[cont_cols].nunique()

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

remove pub_rec from cont and add to cat cols..

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Let's update our continuous and categorical column lists to move `pub_rec` from continuous to categorical features:

```python
# Remove pub_rec from continuous columns
cont_cols = [col for col in cont_cols if col != 'pub_rec']

# Add pub_rec to categorical columns if not already there
if 'pub_rec' not in cat_cols:
    cat_cols.append('pub_rec')
    
# Check the updated lists
print(f"Continuous columns: {len(cont_cols)}")
print(f"Categorical columns: {len(cat_cols)}")
```

In [ ]:
cont_cols = [col for col in cont_cols if col != 'pub_rec']

# Add pub_rec to categorical columns if not already there
if 'pub_rec' not in cat_cols:
    cat_cols.append('pub_rec')
    
# Check the updated lists
print(f"Continuous columns: {len(cont_cols)}")
print(f"Categorical columns: {len(cat_cols)}")

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cont_cols = [col for col in cont_cols if col != 'pub_rec']
      3 # Add pub_rec to categorical columns if not already there
      4 if 'pub_rec' not in cat_cols:

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

In [ ]:
cont_cols

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cont_cols

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

In [ ]:
cat_cols

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cat_cols

NameError: name 'cat_cols' is not defined


NameError: name 'cat_cols' is not defined

make this remove ad add into small fucntion...

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's a small function to adjust column categorization:

```python
def adjust_column_types(cont_cols, cat_cols, move_to_cat=None, move_to_cont=None):
    "Move columns between continuous and categorical lists"
    if move_to_cat:
        cont_cols = [col for col in cont_cols if col not in move_to_cat]
        for col in move_to_cat:
            if col not in cat_cols: cat_cols.append(col)
    
    if move_to_cont:
        cat_cols = [col for col in cat_cols if col not in move_to_cont]
        for col in move_to_cont:
            if col not in cont_cols: cont_cols.append(col)
            
    return cont_cols, cat_cols
```

And here's how to use it:

```python
# Example usage
cont_cols, cat_cols = adjust_column_types(cont_cols, cat_cols, move_to_cat=['pub_rec'])
```

In [ ]:
def adjust_column_types(cont_cols, cat_cols, move_to_cat=None, move_to_cont=None):
    "Move columns between continuous and categorical lists"
    if move_to_cat:
        cont_cols = [col for col in cont_cols if col not in move_to_cat]
        for col in move_to_cat:
            if col not in cat_cols: cat_cols.append(col)
    
    if move_to_cont:
        cat_cols = [col for col in cat_cols if col not in move_to_cont]
        for col in move_to_cont:
            if col not in cont_cols: cont_cols.append(col)
            
    return cont_cols, cat_cols

In [ ]:
cont_cols, cat_cols = adjust_column_types(cont_cols, cat_cols, move_to_cat=['pub_rec'])

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 cont_cols, cat_cols = adjust_column_types(cont_cols, cat_cols, move_to_cat=['pub_rec'])

NameError: name 'cont_cols' is not defined


NameError: name 'cont_cols' is not defined

# combined steps..

In [ ]:
from utils import *


In [ ]:
df = pd.read_csv('Assessment Modeling Dataset.csv')
df= preprocess_data(df)
df= feature_engineering(df)
cont_cols,cat_cols = cont_cat_split(df, max_card=20, dep_var='TARGET')
cont_cols, cat_cols = adjust_column_types(cont_cols, cat_cols, move_to_cat=['pub_rec'])


---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 df = pd.read_csv('Assessment Modeling Dataset.csv')
      2 df= preprocess_data(df)
      3 df= feature_engineering(df)

NameError: name 'pd' is not defined


NameError: name 'pd' is not defined

In [ ]:
df[cont_cols].isna().sum()

annual_inc                0
debt_to_income            0
delinq_2yrs               0
inq_last_6mths            0
mths_since_last_delinq    0
mths_since_last_record    0
open_acc                  0
revol_bal                 0
revol_util                0
total_acc                 0
years_since_first_cr      0
dtype: int64

In [ ]:
df[cont_cols].nunique()

annual_inc                1590
debt_to_income            2471
delinq_2yrs                 10
inq_last_6mths              18
mths_since_last_delinq      91
mths_since_last_record      88
open_acc                    35
revol_bal                 6694
revol_util                1020
total_acc                   73
years_since_first_cr        50
dtype: int64

In [ ]:
df[cat_cols].isna().sum()

emp_length                     0
home_ownership                 0
verification_status            0
pymnt_plan                     0
purpose_cat                    0
addr_state                     0
initial_list_status            0
mths_since_last_major_derog    0
policy_code                    0
has_mths_since_last_delinq     0
has_mths_since_last_record     0
pub_rec_cat                    0
pub_rec                        0
dtype: int64

In [ ]:
df[cat_cols].nunique()

emp_length                     14
home_ownership                  5
verification_status             3
pymnt_plan                      2
purpose_cat                    27
addr_state                     49
initial_list_status             2
mths_since_last_major_derog     3
policy_code                     5
has_mths_since_last_delinq      2
has_mths_since_last_record      2
pub_rec_cat                     4
pub_rec                         4
dtype: int64

Does this data processing step make sense? Do you agree with the data form and all the steps I performed? Do you disagree somewhere? Should I do something more before I start building my model? Any obvious mistakes that don't make sense? Anything that should be categorical but is continuous, or vice versa? Please share your opinion and think carefully.


##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Your data processing looks generally good, but I have a few suggestions for improvement:

1. **Redundant columns**:
   - You have both `pub_rec` and `pub_rec_cat` in your categorical columns - they're redundant. Keep only one (I'd suggest just `pub_rec`).

2. **Additional categorical conversions**:
   - `delinq_2yrs` has only 10 unique values with a very skewed distribution (7120 zeros). This should likely be categorical.
   - `inq_last_6mths` with 18 unique values is borderline - consider testing both ways.

3. **Feature transformations**:
   - `revol_bal` looks heavily skewed based on your describe output (mean vs median). Consider log-transforming it.
   - `annual_inc` might also benefit from log transformation.

4. **Feature binning**:
   - `debt_to_income` could be binned into ranges that represent meaningful risk segments
   - `years_since_first_cr` could be binned into credit history length categories

5. **Additional features**:
   - Consider ratio features like `revol_bal / annual_inc` for debt burden
   - `delinq_2yrs > 0` as a binary flag might be more predictive than the raw count

6. **Feature interactions**:
   - Simple interactions between features like `home_ownership` × `debt_to_income` might reveal important patterns

7. **Column duplication check**:
   - Some columns might be representing the same information (multicollinearity)

Overall, your processing pipeline is logical and handles the basics well. These additional steps could further improve your model's discriminatory power.

Does the data look good enough to start building the first model? We can always come back and do more feature engineering, but overall, do you think it makes sense to build an initial model with this data as is?


##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Yes, the data is definitely ready for an initial model! You've done solid preprocessing work:

- Missing values have been handled appropriately
- High-cardinality categorical columns have been managed
- The date field has been converted to a more useful numeric feature
- Continuous and categorical features have been properly separated

Building an initial baseline model with this data makes perfect sense. This will give you:

1. A performance benchmark to improve upon
2. Feature importance insights to guide further engineering
3. A better understanding of which areas need more focus

You can always iterate and add more sophisticated feature engineering after evaluating your first model. This iterative approach is actually preferable to spending too much time on feature engineering before seeing how a simple model performs.

I'd suggest moving forward with building your first model now, then using what you learn to guide your next steps.

Let's build our tabular pandas. I want to build a fast AI model, and we'll go in from there.


##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Let's build the TabularPandas and model with fastai:

```python
from fastai.tabular.all import *

# Create train/validation splits
splits = RandomSplitter(valid_pct=0.2)(range_of(df))

# Create TabularPandas
to = TabularPandas(df, procs=[Categorify, FillMissing, Normalize],
                  cat_names=cat_cols,
                  cont_names=cont_cols,
                  y_names='TARGET',
                  splits=splits)

# Create DataLoaders
dls = to.dataloaders(bs=64)

# Create tabular learner
learn = tabular_learner(dls, metrics=[accuracy, RocAucBinary()])

# Train the model
learn.fit_one_cycle(5, 1e-2)
```

This will give us a solid baseline model. We're using:
- A 80/20 train/validation split
- Standard preprocessing (categorification, missing value handling, normalization)
- Binary classification metrics (accuracy and ROC AUC)
- The standard 1cycle training policy

After this runs, we can evaluate performance and decide on next steps.

In [ ]:
from fastai.tabular.all import *

# Create train/validation splits
splits = RandomSplitter(valid_pct=0.2)(range_of(df))

# Create TabularPandas
to = TabularPandas(df, procs=[Categorify, FillMissing, Normalize],
                  cat_names=cat_cols,
                  cont_names=cont_cols,
                  y_names='TARGET',
                  splits=splits)

# Create DataLoaders
dls = to.dataloaders(bs=64)

# Create tabular learner
learn = tabular_learner(dls, metrics=[accuracy, RocAucBinary()])

# Train the model
learn

In [ ]:
learn

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 learn

NameError: name 'learn' is not defined


NameError: name 'learn' is not defined

In [ ]:
# fit the Learner
learn.fit_one_cycle(5, 1e-2)


epoch     train_loss  valid_loss  accuracy  roc_auc_score  time    
0         0.143037    0.124170    0.875000  0.500000       00:01     
1         0.112310    0.135443    0.875000  0.500000       00:01     
2         0.101141    0.102670    0.875000  0.500000       00:01     
3         0.090061    0.096437    0.875000  0.500000       00:01     
4         0.085454    0.096734    0.875000  0.500000       00:01     


lets add class weight to handle the unbalanced target class.

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Let's add class weights to handle the imbalanced target class:

```python
# Calculate class weights based on the target distribution
target_counts = df['TARGET'].value_counts()
total = len(df)
weights = torch.FloatTensor([total/target_counts[0], total/target_counts[1]])
weights = weights / weights.sum()

# Create the learner with class weights
learn = tabular_learner(dls, metrics=[accuracy, RocAucBinary()])
learn.loss_func = CrossEntropyLossFlat(weight=weights)

# Train the model with class weights
learn.fit_one_cycle(5, 1e-2)
```

This will adjust the loss function to give more importance to the minority class (defaults), helping the model learn better from the imbalanced data.

In [ ]:
target_counts = df['TARGET'].value_counts()
total = len(df)
weights = torch.FloatTensor([total/target_counts[0], total/target_counts[1]])
weights = weights / weights.sum()

In [ ]:
weights

tensor([0.1277, 0.8723])

In [ ]:
learn = tabular_learner(dls, metrics=[accuracy, RocAucBinary()])
learn.loss_func = CrossEntropyLossFlat(weight=weights)

# Train the model with class weights
learn.fit_one_cycle(5, 1e-2)

epoch     train_loss  valid_loss  accuracy  roc_auc_score  time    
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[1], line 5
      2 learn.loss_func = CrossEntropyLossFlat(weight=weights)
      4 # Train the model with class weights
----> 5 learn.fit_one_cycle(5, 1e-2)

File ~/.local/lib/python3.11/site-packages/fastai/callback/schedule.py:121, in fit_one_cycle(self, n_epoch, lr_max, div, div_final, pct_start, wd, moms, cbs, reset_opt, start_epoch)
    118 lr_max = np.array([h['lr'] for h in self.opt.hypers])
    119 scheds = {'lr': combined_cos(pct_start, lr_max/div, lr_max, lr_max/div_final),
    120           'mom': combined_cos(pct_start, *(self.moms if moms is None else moms))}
--> 121 self.fit(n_epoch, cbs=ParamScheduler(scheds)+L(cbs), reset_opt=reset_opt, wd=wd, start_epoch=start_epoch)

File ~/.local/lib/python3.11/site-packages/fastai/learner.py:272, in Learner.fi

RuntimeError: weight tensor should be defined either for all 1 classes or no classes but got weight tensor of shape: [2]

refer Loss_Functions_fastai_docs docs for context and first tell me intuitively what wrong here why we using this loss fucntion and why it failing with simple example please..

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Looking at the error and the fastai Loss Functions documentation, I can see the issue with our approach.

The error message says: `weight tensor should be defined either for all 1 classes or no classes but got weight tensor of shape: [2]`

The problem is that we're using `CrossEntropyLossFlat` incorrectly. Let me explain:

In fastai, for binary classification tasks like ours (default vs. no default), the model typically outputs a single value which is then passed through a sigmoid function. The target is either 0 or 1. This is fundamentally different from multi-class classification where the model outputs multiple logits (one for each class).

When we look at the Loss_Functions_fastai_docs, we can see that:

1. `CrossEntropyLossFlat` is designed for multi-class classification
2. For binary classification, we should use `BCEWithLogitsLossFlat` instead

Here's a simple example:

```
# Wrong approach (what we did):
# For binary classification with TARGET values of 0 and 1
weights = torch.FloatTensor([0.1277, 0.8723])
loss_func = CrossEntropyLossFlat(weight=weights)  # Error! Expects n_classes logits

# Correct approach:
# For binary classification, we use BCEWithLogitsLossFlat
pos_weight = torch.FloatTensor([0.8723/0.1277])  # Weight for positive class (ratio of neg:pos)
loss_func = BCEWithLogitsLossFlat(pos_weight=pos_weight)
```

The issue is that our model is outputting a single value for each sample (for binary classification), but CrossEntropyLoss expects logits for each class. Let's fix this by using the appropriate loss function for binary classification.

In [ ]:
learn = tabular_learner(dls, metrics=[accuracy, RocAucBinary()])
pos_weight = torch.FloatTensor([0.8723/0.1277])  # Weight for positive class (ratio of neg:pos)
loss_func = BCEWithLogitsLossFlat(pos_weight=pos_weight)

# Train the model with class weights
learn.fit_one_cycle(5, 1e-2)

epoch     train_loss  valid_loss  accuracy  roc_auc_score  time    
0         0.145086    0.121886    0.875000  0.500000       00:01     
1         0.106122    0.111109    0.875000  0.500000       00:01     
2         0.096713    0.101208    0.875000  0.500000       00:01     
3         0.089435    0.096973    0.875000  0.500000       00:01     
4         0.087512    0.098837    0.875000  0.500000       00:01     


let  try the mteric mention in assessment ...

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Please post on Discord that you got this error:
Expecting value: line 2 column 1 (char 1)